From a1427aad7cefbe952c9db37a4a46d9b0af672943 Mon Sep 17 00:00:00 2001
From: Patrick Thier <pthier@chromium.org>
Date: Fri, 15 Oct 2021 15:04:15 +0000
Subject: [PATCH] Assert that we never copy properties from an object itself

When copying properties, it should never happen that source == target.
Add a CHECK to assert this assumption.

Bug: chromium:1260129
Change-Id: Ia5248e4363d85e13052db726fb7143897cea9c87
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/3226779
Commit-Queue: Patrick Thier <pthier@chromium.org>
Reviewed-by: Toon Verwaest <verwaest@chromium.org>
Cr-Commit-Position: refs/heads/main@{#77418}
---

diff --git a/src/objects/js-objects.cc b/src/objects/js-objects.cc
index 26b32ea..dcb7a36 100644
--- a/src/objects/js-objects.cc
+++ b/src/objects/js-objects.cc
@@ -230,6 +230,9 @@
     return Just(false);
   }
 
+  // We should never try to copy properties from an object itself.
+  CHECK_IMPLIES(!use_set, !target.is_identical_to(from));
+
   Handle<DescriptorArray> descriptors(map->instance_descriptors(isolate),
                                       isolate);
 
